### ISSUE TITLE
Not clear which compilers are supported

### ISSUE THREAD

**Issue by eyalroz** [ccache/ccache#355]
The official documentation tells us that "compilers behaving like GCC" are supported. That's good to know, but not enough.

What's missing:
1. (bare minimum) A list - in `README.md` and in the documentation - of supported compilers and known-to-be-supported maximum versions. 
2. A table detailing partial support or caveats w.r.t. the various supported compilers. This would go in the documentation, the website or the repository wiki.
3. Explanation of how to set up the use of ccache with various supported compilers (if it's not exactly the same as with GCC).

**Comment by jrosdahl** [comment]
Thanks for the suggestions! It would be good to have documention on compiler support.

However, there are a couple reasons why such documentation does not exist and maybe won't exist in the future either:

* It's currently unknown exactly which compilers are supported. Finding out would likely require a quite large amount of time and effort. And it would arguably be boring.
* If we make an effort to document the current state, we need to keep it updated in the future – I think that incorrect information would be worse than no information in this case. Keeping track of new compiler releases and figuring out which of their changes affect ccache would be hard (and, again, boring) work. And even if we make a guess that a new compiler version should be supported, some more or less obscure feature could be unsupported for a long time without anybody noticing. We will in practice have to wait for a bug report to know.
* The ccache project has very little resources to spare. If we were more active developers it would make more sense to try keeping this kind of documention up to date.

So to sum it up, you are kind of on your own, which is what the formulation "GCC and compilers that behave similar enough" tries to capture.

But perhaps it would be possible to find some middle ground where we add documentation saying more than just "GCC and similar" but that won't require as much time and effort to maintain. Would you be interested in helping out with this? Do you have any suggestions?

**Comment by eyalroz** [comment]
Well, you could at least say which compilers are supported to the best of your knowledge; and which compilers are known to be used with ccache by a large number of people/organizations. Even with a disclaimer that this does not mean perfect support - that is very important IMHO.

**Comment by jrosdahl** [comment]
> Well, you could at least say which compilers are supported to the best of your knowledge;

Yes, that would be possible to do.

> and which compilers are known to be used with ccache by a large number of people/organizations.

It probably comes as a surprise to you, but I have very little knowledge about that.

> Even with a disclaimer that this does not mean perfect support - that is very important IMHO.

Since you brought up the topic and think that it is very important, could you perhaps describe what you specifically would have liked to be able to tell from the documentation? I.e., which compiler and compiler version in combination with ccache would you like to know about?

**Comment by afbjorklund** [comment]
For myself, "similar" normally means [clang](https://clang.llvm.org/).

We do need help with non-Linux platforms.

**Comment by eyalroz** [comment]
@afbjorklund : There are over 20 different C++ compilers on Unix-like platforms. See [this list](https://en.wikipedia.org/wiki/List_of_compilers#C++_compilers)...

**Comment by eyalroz** [comment]
> Yes, that would be possible to do.

How about starting with that, and some sort of solicitation of users to report successes/failures/configuration instructions with various compilers or platforms? Maybe through a wiki Or a similar collaborative mechanism?

**Comment by afbjorklund** [comment]
@eyalroz : that might be true, but I am mostly looking at gcc and clang myself. There are enough versions of those two...

**Comment by jrosdahl** [comment]
> How about starting with that

Given time and priority, sure. At the moment this does not feel like high priority to me.

> and some sort of solicitation of users to report successes/failures/configuration instructions with various compilers or platforms? Maybe through a wiki Or a similar collaborative mechanism?

Sounds reasonable, but I won't have time to drive this. If you would like to help out I can open up the wiki so that you can start creating some framework for collecting information.

Also, any answer to my [previous questions](https://github.com/ccache/ccache/issues/355#issuecomment-465300037)?

**Comment by jrosdahl** [comment]
I'm working on improving this.

**Comment by jrosdahl** [comment]
@eyalroz: I've now created a _Supported platforms, compilers and languages_ page: https://ccache.dev/platform-compiler-language-support.html. It doesn't fulfill what you thought would be the "bare minimum", though:

* The information is not in the README. This is because I believe that this kind of information does not belong in the README (which is included in each source release) but in a more living document.
* The page does not contain known-to-be-supported maximum compiler versions since a) I don't know them, and b) I don't think that it will be possible to keep that level of detail up to date.
* As mentioned on the page, I welcome contributions of more information on supported and unsupported compilers and platforms.

By the way, I'm still interested in an answer to my previous question in https://github.com/ccache/ccache/issues/355#issuecomment-465300037.

This is the best I can do for now. Closing.

**Comment by eyalroz** [comment]
The new page is awesome! A huge improvement ... and a link from the README is fine.

Exact compiler versions supported are still important, for people who are not  using the latest compiler versions that you're auto-testing with.

> By the way, I'm still interested in an answer to my previous question

I was mostly interested in nVIDIA's NVCC, which is "gcc-like" (and uses gcc) - and the page you've created answers that. (Although I still need to try and get it to work.)

**Comment by eyalroz** [comment]
Oh, one minor nitpick: the CUDA line on the page should mention both `.cu` and `.cuh` files, like `.cpp` and `.hpp` for C++.

**Comment by jrosdahl** [comment]
> Oh, one minor nitpick: the CUDA line on the page should mention both .cu and .cuh files, like .cpp and .hpp for C++.

Right. `.cuh` should actually be on a separate "C-level support" line since it's not supported. Added now.
